keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

```diff
@@ -0,0 +1,359 @@
+
+### Goals
+
+- To allow declarative control of kubeadm action that leads to "in place" mutations[1] of kubeadm generated artifacts. More specifically this initially targets includes kubeadm workflows:
+  - kubeadm upgrade
+  - certificate renewal
+  - change settings/configuration in an existing cluster (NEW) 
+  - certificate authority rotation (NEW)
+- To automate the orchestration of such mutations across the existing nodes in a cluster
+
+> [1] Please note that we are referring to "in place" mutations of kubeadm generated artifacts in order to highlight the difference between the kubeadm operator and other SCL projects like [Cluster API](https://cluster-api.sigs.k8s.io/), which instead assume node/machines immutable
+
+Considering the complexity of this topic, this document is expected to be subject to some iterations. The goal of the current iteration is to:
+
+- Get initial approval on Summary and Motivation paragraphs
+- To identify a semantic for defining “Actions” to be performed by the kubeadm-operator.
+- To define how the kubeadm-operator should manage kubeadm workflows.
+- To define how the users should interact with the kubeadm-operator, including also observability and error handling.
+- To define how the kubeadm-operator should be deployed in a Kubernetes cluster.
+
+### Non-Goals
+
+- To provide or manage any infrastructure elements like e.g. machines, load balancer, etc.
+- To manage/automate the kubeadm init/join workflows.
+- To provide or manage any component/configuration file/artifacts not generated by kubeadm, with the only exception of the kubelet/kubeadm/kubectl binaries which are considering in scope as required by the upgrade workflow.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes operator, I would like to be able to declaratively control upgrades in a systematic fashion.
+
+#### Story 2
+
+As a Kubernetes operator, I would like to be able to declaratively control certificate renewal in a systematic fashion.
+
+#### Story 3
+
+As a Kubernetes operator, I would like to be able to declaratively control changes of the current cluster's settings in a systematic fashion.
+
+#### Story 4
+
+As a Kubernetes operator, I would like to be able to declaratively certificate authority rotation.
+
+#### Story 5 
+As a Kubernetes operator, I would like to control whether nodes are cordoned and drained for tasks or if they are performed "hot"
+
+### Implementation Details/Notes/Constraints
+
+The initial proposal for implementing the kubeadm operator can be summarized by the following sequence diagram; the goal of the current iteration is to validate/improve the proposed approach.
+
+![](20190916-kubeadm-operator.png)
+
+#### Action
+The first step of the sequence above is the user asking the kubeadm operator to perform an `Action`; `Action` is a new CRD that defines actions performed by the kubeadm operator in a high-level, end-user oriented format e.g. 
```

Comment:
Well this is really an interesting question.

When initially thinking about the operator, I was thinking of making the operator reconcile changes on the `ClusterConfiguration` object, assuming this as a description of the "desired state".

However, it quickly came to me that detecting changes to the `ClusterConfiguration` object and mapping such changes to the kubeadm workflows is far from trivial, and TBH, also a huge paradigm shift from what we are doing today. 
Are we, as a kubeadm maintainers, able to figure out all the knobs of this evolution?
Most importantly, are the kubeadm users ready for moving from today's "full control" to "total trust" to the kubeadm operator?

This is why I decided to reduce the scope and focus only on the part of the operator that orchestrates kubeadm workflows, the Actions.

And this is also the reason why I'm assuming that the operator should never automatically delete Action or Task, because outstanding observability of what is happening/what happened is a must for getting the users to trust the operator. 

Just to be clear, as soon as the orchestrates layer of the operator is in place and reliable, I'm totally OK in moving forward and start to design an higher-level controller that automatically generates actions according to the changes to the `ClusterConfiguration` object. 
I just feel we should get there step by step.

Reply:
[1] after giving this more thought, I think we will endup implementing bits of the Jobs + TTL controller if we want cleanup (which alpha IIRC).  I'd prefer us to refine the flows (short-lived, long-lived, short-lived with cleanup etc..) and see

cc @timothysc @neolit123

Reply:
Indeed, I can see all of your points here.
My personal preference is more on the declarative approach (hence actions are a bit uncanny to me). Diffing the kubeadm ClusterConfiguration in its entirety can be tricky and defining an action for each changed field is probably a doomed effort.
What I think, we should try to do is to start with an operator specific `ClusterConfiguration`. That is something on the lines of:
```yaml
apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
kind: ClusterConfiguration
metadata:
     name: my-cluster
kubernetesVersion: 1.20.0
certificateRotationCount: 3
```

As you can see, this is not the kubeadm, but an operator specific `ClusterConfiguration`. It contains only fields, that we know how to tackle should they get changed. Some of the fields may not even be present in the kubeadm `ClusterConfiguration` (in this example, `certificateRotationCount`).
That way we still end with a declarative approach and a narrow set of actions for a start.

In addition to this, I do have the same concern as @yastij, that we may end up re-implementing large chunks of the Job and/or DaemonSet controllers.

Reply:
Ah so a wrapper on top of a cluster configuration? My expectation here is that when this operator is installed it will generate the object that reflects the state of current configuration. Is that right? Then, a modification to this object will result in some actions on the cluster in an attempt to get the cluster to match the supplied configuration.

I think that's a reasonable way to build up to all the features we'd need to support and I would be in favor of this approach.

Reply:
> Ah so a wrapper on top of a cluster configuration?

Depends on what you call "wrapper on top of a cluster configuration". The full `kubeadm.ClusterConfiguration` don't need to be exposed, but only a subset of fields from it. The new `operator.ClusterConfiguration` should also have some fields to deal with concepts, that are not reflected in the `kubeadm.ClusterConfiguration`. Those concept are usually backed by kubeadm actions. Certificate rotation is one such thing.

For example, the above proposition is roughly going to match the following struct:

```golang
type ClusterConfiguration struct {
	TypeMeta
	
	// Here, ObjectMeta.Name is going to correspond to kubeadmapiv1beta2.ClusterConfiguration.ClusterName
	ObjectMeta
	
	// 1:1 correspondence to kubeadmapiv1beta2.ClusterConfiguration.KubernetesVersion
	// A change here is going to trigger upgrade on the machines to the target version
	KubernetesVersion string
	
	// A new field, not present in kubeadmapiv1beta2.ClusterConfiguration.
	// This is a number, that can only be bumped up. A change to this will trigger certificate rotation.
	CertificateRotationCound int
}
```

> My expectation here is that when this operator is installed it will generate the object that reflects the state of current configuration. Is that right? Then, a modification to this object will result in some actions on the cluster in an attempt to get the cluster to match the supplied configuration.

That's my vision of where things should be going too.

Reply:
If we're tossing out ideas, I know this has no relevance to the PR anymore, but it would be cool if the object had a field like `CertificateExpiration` and showed you when the certificates were going to expire. Then you could set the field to a future date and your certs would rotate to get a new expiration date that matched your desired expiration date, perhaps with an "autoRenew: bool" field on that always never let your certificates expire.

I recognize these behaviors can all be expressed with `Action`s, however.

Reply:
@chuckha @rosti @yastij 
Point taken.
Please consider that we are approaching this incrementally, and mi suggestion is to initially focus on the orchestration layer. In the meantime, we can discuss how to best approach the higher level reconciliation loop, that will be implemented in next iterations.